src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientMethodsAnalyzer.cs

```diff
@@ -184,120 +193,152 @@ private static void CheckProtocolMethodParameters(ISymbolAnalysisContext context
-                    {
-                        context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
-                    }
+                    context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
                 }
-
                 return;
             }
 
             context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0018, method.Locations.FirstOrDefault()), method);
         }
 
-        private static void CheckClientMethodReturnType(ISymbolAnalysisContext context, IMethodSymbol method)
+        private static bool IsOrImplements(ITypeSymbol typeSymbol, string typeName)
         {
-            bool IsOrImplements(ITypeSymbol typeSymbol, string typeName)
+            if (typeSymbol.Name == typeName)
             {
-                if (typeSymbol.Name == typeName)
-                {
-                    return true;
-                }
-
-                if (typeSymbol.BaseType != null)
-                {
-                    return IsOrImplements(typeSymbol.BaseType, typeName);
-                }
+                return true;
+            }
 
-                return false;
+            if (typeSymbol.BaseType != null)
+            {
+                return IsOrImplements(typeSymbol.BaseType, typeName);
             }
 
+            return false;
+        }
+
+        private static void CheckClientMethodReturnType(ISymbolAnalysisContext context, IMethodSymbol method)
+        {
             ITypeSymbol originalType = method.ReturnType;
             ITypeSymbol unwrappedType = method.ReturnType;
 
             if (method.ReturnType is INamedTypeSymbol namedTypeSymbol &&
                 namedTypeSymbol.IsGenericType &&
-                namedTypeSymbol.Name == "Task")
+                namedTypeSymbol.Name == TaskTypeName)
             {
                 unwrappedType = namedTypeSymbol.TypeArguments.Single();
             }
 
-            if (IsOrImplements(unwrappedType, "Response") ||
-                IsOrImplements(unwrappedType, "NullableResponse") ||
-                IsOrImplements(unwrappedType, "Operation") ||
-                IsOrImplements(originalType, "Pageable") ||
-                IsOrImplements(originalType, "AsyncPageable") ||
+            if (IsOrImplements(unwrappedType, ResponseTypeName) ||
+                IsOrImplements(unwrappedType, NullableResponseTypeName) ||
+                IsOrImplements(unwrappedType, OperationTypeName) ||
+                IsOrImplements(originalType, PageableTypeName) ||
+                IsOrImplements(originalType, AsyncPageableTypeName) ||
                 originalType.Name.EndsWith(ClientSuffix))
             {
                 return;
             }
 
             context.ReportDiagnostic(Diagnostic.Create(Descriptors.AZC0015, method.Locations.FirstOrDefault(), originalType.ToDisplayString()), method);
-
         }
 
+        private bool IsCheckExempt(IMethodSymbol method)
```

Comment:
Curious: How did we decide on this logic?  It makes me a little nervous to start exempting a large group of methods from analysis checks.  How do we know we won't accidentally exempt something we previously checked?

Reply:
Ordinary: means it is not a constructor, not a property, etc

OverriddenMethod: below is an example. Think if service method is an overridden method, then it means client should inherited from others. I never see such case. So I'm just exempting the overridden methods from `object`.
![image](https://github.com/Azure/azure-sdk-tools/assets/97225342/2f32077f-dffb-40d6-bd5b-13cba0223ecc)

IsImplicitlyDeclared: it was added before and I don't want to touch it. From comments I believe it is used to exempt the methods automatically created by the compiler, for example, default constructor.